### DISCUSSION TITLE
Email templates not working

### DISCUSSION THREAD

**Discussion by TD-4242** [magicbell/community#145] [Questions] [answered]
I have my general email template setup with a header and footer plus a {{ template.content }} portion.  this looks and seems right:

![image](https://github.com/magicbell-io/community/assets/3487671/dcf28582-79cd-42e1-94b4-f2ad363ae5ba)

but when I go to the categories to fill in the content I put in the template data like so:

![image](https://github.com/magicbell-io/community/assets/3487671/7b45fdac-56eb-487c-a273-82900454fd4e)

when I hit save, it does something to the html although the template still mostly seems to render
![image](https://github.com/magicbell-io/community/assets/3487671/01d35b90-b6d5-4f08-b089-3a302204019c)

although only one of my template vars is working.  If I only put the template vars they show up ok
![image](https://github.com/magicbell-io/community/assets/3487671/4bcf06f6-d071-4a41-b1d2-c2c21f15ceb4)

Also the Category templates use the last notification broadcast no matter the category which makes testing a bit of a pain.

**Comment by unamashana** [comment]
Hey @TD-4242 thank you for the detailed report. We will investigate and get back to you.

**Reply by TD-4242** [reply]
Any updates on this?  I will have to find an alternative if I can't get this working soon.

**Reply by unamashana** [reply]
@TD-4242 unfortunately, we are finding some issues with templates and nested custom_attributes. We are rewriting that bit, but we'll need a couple weeks to get it out. In the meantime, you can send us rendered content in an email/provider override as a workaround. I am sorry we don't have more encouraging news at the moment.

**Reply by TD-4242** [reply]
Any updates on this?  I have a work around but I don't really like it since it has hard coded templates being injected into the notifications and send time that I'd like to use the MB UI to edit.

**Reply by unamashana** [reply]
Hey @TD-4242 we are reworking this feature to offer more logs and a tighter validation step for templates. We don't have a ETA to share, but it will be this quarter.

**Reply by sakopov** [reply]
Any updates on this? Seems like a very fundamental issue for email templating. Surprised it hasn't been given a higher priority. This pretty much makes email portion of MagicBell very difficult to maintain unless you want HTML spaghetti code in all clients that send notifications.

**Reply by TD-4242** [reply]
yes, we have a workaround as @sakopov has mentioned but it makes updating and changing difficult.

**Reply by tf0054** [reply]
Hi, I found this problem yesterday and also think its critical but should I understand the template provided by the smtp provider like Sendgrid still can be used?

https://www.magicbell.com/docs/channels/email/sendgrid#example-sendgrid-email-templates

And if I will use that way, template definded on the provider can be choosen freely bc its option when sending each notifications?

**Reply by unamashana** [reply, marked as answer]
Hey @TD-4242 and @tf0054 - sorry about the delay in getting back to you with a solution. We've been rolling out a new architecture with a new API for uploading templates. We have this up on a review env if you'd like to try it out. I can send you the link and docs if you ping us on Slack - magicbell.to/slack. We are hoping to take this to production in a couple of weeks and actively seeking feedback from customers like yourself who have found the current feature lacking (correctly so!). I am confident you'd be very pleased with the new APIs that offer a robust code as infra style workflow.

**Reply by TD-4242** [reply]
I am back to trying to add a template using a simple SMS template:

```
Your scan "{{ notification.title }}"

{{ notification.content }}

https://symbolik.com/scan-results/?id={{ custom_attributes.actions[0].data.layoutState.showId }}
```

I seem to be unable to get it to work.

**Reply by TD-4242** [reply]
I've added some simplification to the custom_attributes in order to try to make this work a little better.  My current SMS template is much simpler than it was:
```
Your scan "{{ notification.title }}"

{{ notification.content }}

https://symbolik.com/sr/{{ notification.custom_attributes.scan_results_id }} 
```
but it still doesn't seem to work.  I've even just tried:
`{{ notification.custom_attributes.scan_results_id }}`
and
`{{ broadcast.custom_attributes.scan_results_id }}`
and
`{{ custom_attributes.scan_results_id }}`
and the custom attribute will not get populated

even though the `scan_results_id` is in the notification:
```
{
  "broadcast": {
    "id": "018ee787-89b6-7021-bd8f-a1d705bf3b9f",
    "title": "scanScheduleExample",
    "custom_attributes": {
      "icon": { "name": "scan-notification", "type": "system_svg" },
      "toast": { "duration": 5000},
      "actions": [
        {
          "data": {
            "moduleId": "scanGroup",
            "layoutState": {
              "scanId": "cl3dit84o4bqb0u339e183q8d",
              "showId": "bdvlubvt102iiyawusdcezq4",
              "resultSetId": "bdvlubvt102iiyawusdcezq4",
              "showResourceId": "cl3dit84o4bqb0u339e183q8d"
            },
            "forceOpenPanel": true,
            "scanResultsCount": "1632"
          },
          "type": "customEvent",
          "title": "View",
          "value": "NOTIFICATION_MFE:LAYOUTSTATE_SET"
        }
      ],
      "scan_results_id": "bdvlubvt102iiyawusdcezq4",
      "scan_results_count": "1632"
    },
    "content": "Completed: 1632 Results",
    "category": "Scans",
    "topic": "scan-cl3dit84o4bqb0u339e183q8d",
    "overrides": {
      "providers": {
        "amazon_ses": {
          "Source": "Symbolik <noreply@symbolik.com>",
          "Message.Subject.Data": "Your Symbolik Scan Results",
          "Message.Body.Html.Data": "<!DOCTYPE html  ... SES template ... >"
        }
      }
    },
    "recipients": [{ "external_id": "ckv2wg6ro07z00tlpatow8ofb" }],
    "status": {
      "errors": [],
      "status": "processed",
      "summary": { "total": 1, "failures": 0}
    },
    "sent_at": "2024-04-16T15:30:03.318531Z",
    "created_at": "2024-04-16T15:30:03.318531Z"
  }
}
```

**Reply by TD-4242** [reply]
tried narrowing down the other way, and just using `{{ notification.custom_attributes }}` and I can see the full shape of the object and what's needed it's just not working

![image](https://github.com/magicbell-io/community/assets/3487671/978dc0e0-2ac2-4978-a550-ba5ee2556266)

**Reply by TD-4242** [reply]
after playing around with Liquid syntax I think I finally have a solution:
```
Your scan "{{ notification.title }}"

{{ notification.content }}

https://symbolik.com/sr/{{ notification.custom_attributes["scan_results_id"] }}
```
Although this still doesn't show up right in the Content section of the UI.  the API call does end up with the correct text though

![image](https://github.com/magicbell-io/community/assets/3487671/4fd60878-1d48-442b-9ada-a30f78db077b)
